tests/chainer_tests/functions_tests/array_tests/test_copy.py

```diff
@@ -18,97 +18,184 @@ def _to_gpu(x, device_id):
         return x
 
 
-@testing.parameterize(*testing.product({
-    'dtype': [numpy.float16, numpy.float32, numpy.float64],
-}))
-class TestCopy(unittest.TestCase):
+_nonchainerx_backend_configs = (
+    [
+        # NumPy
+        {},
+        # CuPy
+        {'use_cuda': True, 'cuda_device': 0},
+        {'use_cuda': True, 'cuda_device': 1},
+    ])
+
+
+_chainerx_backend_configs = (
+    [
+        {'use_chainerx': True, 'chainerx_device': 'native:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:0'},
+        {'use_chainerx': True, 'chainerx_device': 'cuda:1'},
+    ])
+
+_numpy_device = chainer.get_device('@numpy')
+
+
+class CopyTestBase(object):
 
     def setUp(self):
-        self.x_data = numpy.random.uniform(
-            -1, 1, (10, 5)).astype(self.dtype)
+        self.x = numpy.random.uniform(-1, 1, (10, 5)).astype(self.dtype)
         self.gy = numpy.random.uniform(-1, 1, (10, 5)).astype(self.dtype)
         self.ggx = numpy.random.uniform(-1, 1, (10, 5)).astype(self.dtype)
         self.check_double_backward_options = {}
         if self.dtype == numpy.float16:
             self.check_double_backward_options = {'atol': 5e-3, 'rtol': 5e-2}
 
-    def check_forward(self, src_id, dst_id):
-        x_data = _to_gpu(self.x_data, src_id)
-        x = chainer.Variable(x_data)
-        y = functions.copy(x, dst_id)
+    def check_forward(self, dst_device_spec, src_device, dst_device):
+        x = src_device.send(self.x)
+
+        x_var = chainer.Variable(x)
+        y = functions.copy(x_var, dst_device_spec)
+
+        assert y.device == dst_device
+        assert backend.get_device_from_array(y.array) == dst_device
+        assert y.dtype == self.dtype
+        numpy.testing.assert_array_equal(_numpy_device.send(y.array), self.x)
+
+    def test_forward(self, src_backend_config, dst_backend_config):
+        self.check_forward(
+            dst_backend_config.device,
+            src_backend_config.device,
+            dst_backend_config.device)
+
+    def test_backward(self, src_backend_config, dst_backend_config):
+        x = src_backend_config.get_array(self.x)
+        gy = dst_backend_config.get_array(self.gy)
+        src_device = src_backend_config.device
+        dst_device = dst_backend_config.device
+
+        x_var = chainer.Variable(x)
+
+        y_var = functions.copy(x_var, dst_device)
+        y_var.grad = gy
+
+        y_var.backward()
+
+        x_grad = x_var.grad
+        assert x_var.grad_var.device == src_device
+        assert backend.get_device_from_array(x_grad) == src_device
+        numpy.testing.assert_array_equal(_numpy_device.send(x_grad), self.gy)
+
+    def test_double_backward(self, src_backend_config, dst_backend_config):
+        x = src_backend_config.get_array(self.x)
+        gy = dst_backend_config.get_array(self.gy)
+        ggx = src_backend_config.get_array(self.ggx)
+        dst_device = dst_backend_config.device
+
+        x_var = chainer.Variable(x, requires_grad=True)
```

Comment:
Do we need the `requires_grad=True`? I think it increases the reader's load a bit since there's no such argument passed in `test_backward`.

Reply:
`requires_grad=True` is essential in this test and it's better to make it explicit.

Reply:
Is it not the case for `test_backward` ?

Reply:
It is.